File: lib/rgfa/field/identifier_gfa2.rb
module RGFA::Field::IdentifierGFA2

  def unsafe_decode(string)
    string.to_sym
  end

  def decode(string)
    validate_encoded(string)
    string.to_sym
  end

  def validate_encoded(string)
    if string !~ /^[!-~]+$/
      raise RGFA::FormatError,
        "#{string.inspect} is not a valid GFA2 identifier\n"+
        "(it contains spaces or non-printable characters)"
    end
  end

  def validate_decoded(object)
    case object
    when RGFA::Line
      validate_encoded(object.name)
    when String, Symbol
      validate_encoded(object)
    else
      raise RGFA::TypeError,
        "the class #{object.class} is incompatible with the datatype\n"+
        "(accepted classes: String, Symbol, RGFA::Line)"
    end
  end

  def unsafe_encode(object)
    case object
    when String
      return object
    when Symbol
      return object.to_s
    when RGFA::Line
      return object.name.to_s
    else
      raise RGFA::TypeError,
        "the class #{object.class} is incompatible with the datatype\n"+
        "(accepted classes: String, Symbol, RGFA::Line)"
    end
  end

  def encode(object)
    string = unsafe_encode(object)
    validate_encoded(string)
    return string
  end

  module_function :decode
  module_function :unsafe_decode
  module_function :validate_encoded
  module_function :validate_decoded
  module_function :unsafe_encode
  module_function :encode

end

File: gfapy/field/identifier_gfa2.py
import gfapy
import re

def unsafe_decode(string):
  return string

def decode(string):
  validate_encoded(string)
  return string

def validate_encoded(string):
  if not re.match("^[!-~]+$", string):
    raise gfapy.FormatError(
      "{} is not a valid GFA2 identifier\n".format(repr(string))+
      "(it contains spaces or non-printable characters)")

def validate_decoded(obj):
  if isinstance(obj, gfapy.Line):
    validate_encoded(obj.name)
  elif isinstance(obj, str):
    validate_encoded(obj)
  else:
    raise gfapy.TypeError(
      "the class {} is incompatible with the datatype\n"
      .format(obj.__class__.__name__)+
      "(accepted classes: str, gfapy.Line)")

def unsafe_encode(obj):
  if isinstance(obj, str):
    return obj
  if isinstance(obj, gfapy.Line):
    return str(obj.name)
  else:
    raise gfapy.TypeError(
      "the class {} is incompatible with the datatype\n"
      .format(obj.__class__.__name__)+
      "(accepted classes: str, gfapy.Line)")

def encode(obj):
  string = unsafe_encode(obj)
  validate_encoded(string)
  return string
